daemon/policy.go

```diff
@@ -207,20 +208,38 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
 
 	d.policy.Mutex.Lock()
 	if opts != nil {
+		// removedPrefixes tracks prefixes that we are replacing in the rules. We
+		// remove them unconditionally here, but have added them above already, so
+		// the refcount is still correct. The policy repository is locked during
+		// both operations.
+		var removedPrefixes []*net.IPNet
 		if opts.Replace {
 			for _, r := range rules {
-				tmp := d.policy.SearchRLocked(r.Labels)
-				if len(tmp) > 0 {
+				oldRules := d.policy.SearchRLocked(r.Labels)
+				removedPrefixes = append(removedPrefixes, policy.GetCIDRPrefixes(oldRules)...)
+				if len(oldRules) > 0 {
 					d.policy.DeleteByLabelsLocked(r.Labels)
 				}
 			}
 		}
 		if len(opts.ReplaceWithLabels) > 0 {
-			tmp := d.policy.SearchRLocked(opts.ReplaceWithLabels)
-			if len(tmp) > 0 {
+			oldRules := d.policy.SearchRLocked(opts.ReplaceWithLabels)
+			removedPrefixes = append(removedPrefixes, policy.GetCIDRPrefixes(oldRules)...)
+			if len(oldRules) > 0 {
 				d.policy.DeleteByLabelsLocked(opts.ReplaceWithLabels)
 			}
 		}
+
+		if err := ipcache.ReleaseCIDRs(removedPrefixes); err != nil {
+			_ = d.prefixLengths.Delete(prefixes)
+			metrics.PolicyImportErrors.Inc()
+			log.WithError(err).WithField("prefixes", prefixes).Warn(
+				"Failed to release CIDRs in replaced rules during policy add with replace")
+			return d.policy.GetRevision(), err
```

Comment:
If we're going to error out in this case, then I think we also need to `ipcache.ReleaseCIDRs(prefixes)` which is allocated immediately above this opts != nil case and assumes there are no subsequent failures.

I also note that there is a `d.Policy.Mutex.Lock()` immediately above this, but presumably from within this code we could do a release of CIDRs through to the KVstore and potentially hold that lock for a very long time. Do you think that there's something we can do about that?

Reply:
Good points both. I moved the new release handling to after the the section where we hold the policy repo lock. I don't know why I thought it was important to hold it for both operations, but looking at it again I think it's ok to do the releasing later. We could actually defer this but I thought that would create bugs.